File: plugins/youtube-expert/skills/youtube-analysis/scripts/fetch_transcript.py
"""Fetch YouTube video transcript and metadata.

Standalone CLI tool. Outputs JSON to stdout, status messages to stderr.
Primary: youtube-transcript-api (no API key). Fallback: yt-dlp subtitles.
"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from typing import Any

from utils import parse_youtube_url


def _log(msg: str) -> None:
    """Print status message to stderr."""
    print(msg, file=sys.stderr)


def _run_command(cmd: list[str], timeout: int = 60) -> subprocess.CompletedProcess[str]:
    """Run a subprocess command and normalize dependency/time-out failures."""
    try:
        return subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)
    except FileNotFoundError as exc:
        raise RuntimeError(f"Command not found: {cmd[0]}") from exc
    except subprocess.TimeoutExpired as exc:
        raise RuntimeError(f"Command timed out: {cmd[0]}") from exc


def fetch_transcript_api(
    video_id: str,
    lang: str,
) -> tuple[list[dict[str, Any]], str, str]:
    """Fetch transcript via youtube-transcript-api.

    Args:
        video_id: YouTube video ID.
        lang: Preferred language code.

    Returns:
        Tuple of (transcript_segments, language_used, source_name).

    Raises:
        RuntimeError: If transcript unavailable or library fails.
    """
    from youtube_transcript_api import YouTubeTranscriptApi

    _log(f"Fetching transcript via youtube-transcript-api (lang={lang})...")

    ytt = YouTubeTranscriptApi()

    # Try direct fetch with requested language first
    try:
        fetched = ytt.fetch(video_id, languages=[lang])
        segments = [
            {"start": s.start, "duration": s.duration, "text": s.text}
            for s in fetched
        ]
        return segments, lang, "youtube-transcript-api"
    except Exception:
        pass

    # Fall back: list available transcripts, pick first available
    try:
        transcript_list = ytt.list(video_id)
        available = list(transcript_list)
        if not available:
            raise RuntimeError("No transcripts available for this video")
        transcript = available[0]
        _log(f"Requested lang '{lang}' unavailable. Using '{transcript.language_code}'")
        fetched = transcript.fetch()
        segments = [
            {"start": s.start, "duration": s.duration, "text": s.text}
            for s in fetched
        ]
        return segments, transcript.language_code, "youtube-transcript-api"
    except RuntimeError:
        raise
    except Exception as exc:
        raise RuntimeError(f"No usable transcript found: {exc}") from exc


def fetch_transcript_ytdlp(video_id: str, lang: str) -> tuple[list[dict], str, str]:
    """Fetch transcript via yt-dlp subtitle extraction.

    Args:
        video_id: YouTube video ID.
        lang: Preferred language code.

    Returns:
        Tuple of (transcript_segments, language_used, source_name).

    Raises:
        RuntimeError: If yt-dlp fails or no subtitles found.
    """
    import tempfile
    from pathlib import Path

    _log("Falling back to yt-dlp for transcript...")
    url = f"https://www.youtube.com/watch?v={video_id}"

    with tempfile.TemporaryDirectory() as tmpdir:
        def build_cmd(language: str | None) -> list[str]:
            cmd = [
                "yt-dlp",
                "--write-sub",
                "--write-auto-sub",
                "--sub-format",
                "json3",
                "--skip-download",
                "--output",
                f"{tmpdir}/%(id)s",
            ]
            if language:
                cmd.extend(["--sub-lang", language])
            cmd.append(url)
            return cmd

        def parse_json3(path: Path) -> list[dict[str, Any]]:
            sub_data = json.loads(path.read_text(encoding="utf-8"))
            events = sub_data.get("events", [])
            segments: list[dict[str, Any]] = []
            for event in events:
                segs = event.get("segs")
                if not segs:
                    continue
                text = "".join(segment.get("utf8", "") for segment in segs).strip()
                if not text or text == "\n":
                    continue
                start_ms = event.get("tStartMs", 0)
                duration_ms = event.get("dDurationMs", 0)
                segments.append(
                    {
                        "start": start_ms / 1000.0,
                        "duration": duration_ms / 1000.0,
                        "text": text,
                    }
                )
            return segments

        def extract_language(path: Path) -> str:
            parts = path.name.rsplit(".", 2)
            if len(parts) >= 3:
                return parts[-2]
            return lang

        def try_extract(language: str | None) -> tuple[list[dict[str, Any]], str] | None:
            result = _run_command(build_cmd(language))
            if result.returncode != 0:
                detail = result.stderr.strip() or result.stdout.strip()
                raise RuntimeError(f"yt-dlp subtitle extraction failed: {detail}")

            sub_files = sorted(Path(tmpdir).glob("*.json3"))
            if not sub_files:
                return None

            preferred = next(
                (
                    path
                    for path in sub_files
                    if language and extract_language(path).startswith(language)
                ),
                sub_files[0],
            )
            segments = parse_json3(preferred)
            if not segments:
                return None
            return segments, extract_language(preferred)

        try:
            extracted = try_extract(lang)
            if extracted is None:
                extracted = try_extract(None)
        except RuntimeError as exc:
            raise RuntimeError(str(exc)) from exc

        if extracted is None:
            raise RuntimeError("yt-dlp produced no usable subtitle files")

        segments, actual_lang = extracted
        if actual_lang != lang:
            _log(f"yt-dlp requested lang '{lang}' unavailable. Using '{actual_lang}'")
        return segments, actual_lang, "yt-dlp"


def fetch_metadata(video_id: str) -> dict[str, Any]:
    """Extract video metadata via yt-dlp.

    Args:
        video_id: YouTube video ID.

    Returns:
        Dict with title, channel, duration_seconds, upload_date,
        description, view_count, tags.
    """
    _log("Fetching metadata via yt-dlp...")
    url = f"https://www.youtube.com/watch?v={video_id}"
    cmd = ["yt-dlp", "--dump-json", "--no-download", url]

    try:
        result = _run_command(cmd)
    except RuntimeError as exc:
        _log(f"Metadata fetch failed: {exc}")
        return {
            "title": "Unknown",
            "channel": "Unknown",
            "duration_seconds": 0,
            "upload_date": "Unknown",
            "description": "",
            "view_count": 0,
            "tags": [],
        }

    if result.returncode != 0:
        _log(f"Metadata fetch failed: {result.stderr.strip() or result.stdout.strip()}")
        return {
            "title": "Unknown",
            "channel": "Unknown",
            "duration_seconds": 0,
            "upload_date": "Unknown",
            "description": "",
            "view_count": 0,
            "tags": [],
        }

    try:
        data = json.loads(result.stdout)
    except json.JSONDecodeError as exc:
        _log(f"Metadata fetch returned invalid JSON: {exc}")
        return {
            "title": "Unknown",
            "channel": "Unknown",
            "duration_seconds": 0,
            "upload_date": "Unknown",
            "description": "",
            "view_count": 0,
            "tags": [],
        }
    upload_date = data.get("upload_date", "")
    if len(upload_date) == 8:
        upload_date = f"{upload_date[:4]}-{upload_date[4:6]}-{upload_date[6:]}"

    return {
        "title": data.get("title", "Unknown"),
        "channel": data.get("channel", data.get("uploader", "Unknown")),
        "duration_seconds": data.get("duration", 0),
        "upload_date": upload_date,
        "description": data.get("description", ""),
        "view_count": data.get("view_count", 0),
        "tags": data.get("tags") or [],
    }


def fetch_video(url_or_id: str, lang: str = "en") -> dict[str, Any]:
    """Fetch transcript and metadata for a YouTube video.

    Args:
        url_or_id: YouTube URL or video ID.
        lang: Preferred transcript language.

    Returns:
        Complete video data dict ready for JSON serialization.

    Raises:
        SystemExit: On unrecoverable errors.
    """
    video_id = parse_youtube_url(url_or_id)
    if not video_id:
        _log(f"ERROR: Cannot parse YouTube URL or video ID: {url_or_id}")
        sys.exit(1)

    _log(f"Video ID: {video_id}")

    # Fetch transcript: primary then fallback
    segments: list[dict] = []
    language = lang
    source = ""

    try:
        segments, language, source = fetch_transcript_api(video_id, lang)
    except (RuntimeError, ImportError) as exc:
        _log(f"Primary transcript fetch failed: {exc}")
        try:
            segments, language, source = fetch_transcript_ytdlp(video_id, lang)
        except RuntimeError as exc2:
            _log(f"Fallback transcript fetch failed: {exc2}")
            _log("ERROR: No transcript available for this video")
            sys.exit(2)

    _log(f"Transcript: {len(segments)} segments via {source} (lang={language})")

    # Fetch metadata
    metadata = fetch_metadata(video_id)

    transcript_text = " ".join(seg["text"] for seg in segments)

    return {
        "video_id": video_id,
        "title": metadata["title"],
        "channel": metadata["channel"],
        "duration_seconds": metadata["duration_seconds"],
        "upload_date": metadata["upload_date"],
        "description": metadata["description"],
        "view_count": metadata["view_count"],
        "tags": metadata["tags"],
        "transcript": segments,
        "transcript_text": transcript_text,
        "language": language,
        "source": source,
    }


def main() -> None:
    """CLI entry point."""
    parser = argparse.ArgumentParser(
        description="Fetch YouTube video transcript and metadata."
    )
    parser.add_argument("url", help="YouTube URL or video ID")
    parser.add_argument(
        "--lang",
        default="en",
        help="Preferred transcript language (default: en)",
    )
    args = parser.parse_args()

    data = fetch_video(args.url, args.lang)
    json.dump(data, sys.stdout, ensure_ascii=False, indent=2)
    print()
    print()  # trailing newline


if __name__ == "__main__":
    main()

File: src/components/scripts/sources/youtube-analysis/fetch_transcript.ts
#!/usr/bin/env node

import { spawnSync } from "node:child_process";
import { mkdtempSync, readdirSync, readFileSync, rmSync, realpathSync } from "node:fs";
import { tmpdir } from "node:os";
import { basename, join } from "node:path";
import { fileURLToPath } from "node:url";
import { parseYoutubeUrl } from "./utils";

function log(message) {
  console.error(message);
}

export function runCommand(command, args, { timeoutMs = 60000 } = {}) {
  const result = spawnSync(command, args, { encoding: "utf8", timeout: timeoutMs });
  if (result.error) {
    if (result.error.code === "ENOENT") throw new Error(`Command not found: ${command}`);
    if (result.error.code === "ETIMEDOUT") throw new Error(`Command timed out: ${command}`);
    throw result.error;
  }
  return result;
}

export function buildSubtitleCommand(videoId, language, outputTemplate) {
  const args = [
    "--write-sub",
    "--write-auto-sub",
    "--sub-format",
    "json3",
    "--skip-download",
    "--output",
    outputTemplate,
  ];
  if (language) {
    args.push("--sub-lang", language);
  }
  args.push(`https://www.youtube.com/watch?v=${videoId}`);
  return args;
}

export function parseJson3(path) {
  const data = JSON.parse(readFileSync(path, "utf8"));
  const segments = [];

  for (const event of data.events || []) {
    if (!Array.isArray(event.segs)) continue;
    const text = event.segs.map((segment) => segment.utf8 || "").join("").trim();
    if (!text || text === "\n") continue;
    segments.push({
      start: Number(event.tStartMs || 0) / 1000.0,
      duration: Number(event.dDurationMs || 0) / 1000.0,
      text,
    });
  }

  return segments;
}

export function extractLanguage(path, fallback = "en") {
  const parts = basename(path).split(".");
  return parts.length >= 3 ? parts[parts.length - 2] : fallback;
}

export function fetchTranscriptYtdlp(videoId, lang = "en", { runner = runCommand, tempDir = null } = {}) {
  log("Fetching transcript via yt-dlp...");
  const workDir = tempDir || mkdtempSync(join(tmpdir(), "youtube-transcript-"));
  const cleanup = !tempDir;

  try {
    const tryExtract = (language) => {
      const result = runner("yt-dlp", buildSubtitleCommand(videoId, language, join(workDir, "%(id)s")));
      if (result.status !== 0) {
        const detail = String(result.stderr || result.stdout || "").trim();
        throw new Error(`yt-dlp subtitle extraction failed: ${detail}`);
      }

      const subtitleFiles = readdirSync(workDir)
        .filter((name) => name.endsWith(".json3"))
        .map((name) => join(workDir, name))
        .sort();
      if (!subtitleFiles.length) return null;

      const preferred = subtitleFiles.find((path) => language && extractLanguage(path, lang).startsWith(language)) || subtitleFiles[0];
      const segments = parseJson3(preferred);
      if (!segments.length) return null;
      return [segments, extractLanguage(preferred, lang)];
    };

    let extracted = tryExtract(lang);
    if (extracted === null) extracted = tryExtract(null);
    if (extracted === null) throw new Error("yt-dlp produced no usable subtitle files");

    const [segments, actualLang] = extracted;
    if (actualLang !== lang) {
      log(`yt-dlp requested lang '${lang}' unavailable. Using '${actualLang}'`);
    }
    return [segments, actualLang, "yt-dlp"];
  } finally {
    if (cleanup) rmSync(workDir, { recursive: true, force: true });
  }
}

function defaultMetadata() {
  return {
    title: "Unknown",
    channel: "Unknown",
    duration_seconds: 0,
    upload_date: "Unknown",
    description: "",
    view_count: 0,
    tags: [],
  };
}

export function fetchMetadata(videoId, { runner = runCommand } = {}) {
  log("Fetching metadata via yt-dlp...");
  const url = `https://www.youtube.com/watch?v=${videoId}`;

  let result;
  try {
    result = runner("yt-dlp", ["--dump-json", "--no-download", url]);
  } catch (error) {
    log(`Metadata fetch failed: ${error.message}`);
    return defaultMetadata();
  }

  if (result.status !== 0) {
    log(`Metadata fetch failed: ${String(result.stderr || result.stdout || "").trim()}`);
    return defaultMetadata();
  }

  try {
    const data = JSON.parse(result.stdout);
    let uploadDate = data.upload_date || "";
    if (/^\d{8}$/.test(uploadDate)) {
      uploadDate = `${uploadDate.slice(0, 4)}-${uploadDate.slice(4, 6)}-${uploadDate.slice(6)}`;
    }
    return {
      title: data.title || "Unknown",
      channel: data.channel || data.uploader || "Unknown",
      duration_seconds: data.duration || 0,
      upload_date: uploadDate,
      description: data.description || "",
      view_count: data.view_count || 0,
      tags: data.tags || [],
    };
  } catch (error) {
    log(`Metadata fetch returned invalid JSON: ${error.message}`);
    return defaultMetadata();
  }
}

export function fetchVideo(urlOrId, lang = "en", options = {}) {
  const videoId = parseYoutubeUrl(urlOrId);
  if (!videoId) {
    const error = new Error(`Cannot parse YouTube URL or video ID: ${urlOrId}`);
    error.exitCode = 1;
    throw error;
  }

  log(`Video ID: ${videoId}`);

  let segments;
  let language;
  let source;
  try {
    [segments, language, source] = fetchTranscriptYtdlp(videoId, lang, options);
  } catch (error) {
    log(`Transcript fetch failed: ${error.message}`);
    const wrapped = new Error("No transcript available for this video");
    wrapped.exitCode = 2;
    throw wrapped;
  }

  log(`Transcript: ${segments.length} segments via ${source} (lang=${language})`);
  const metadata = fetchMetadata(videoId, options);
  const transcriptText = segments.map((segment) => segment.text).join(" ");

  return {
    video_id: videoId,
    title: metadata.title,
    channel: metadata.channel,
    duration_seconds: metadata.duration_seconds,
    upload_date: metadata.upload_date,
    description: metadata.description,
    view_count: metadata.view_count,
    tags: metadata.tags,
    transcript: segments,
    transcript_text: transcriptText,
    language,
    source,
  };
}

export function usage() {
  return `Fetch YouTube video transcript and metadata.

Usage: node scripts/fetch_transcript.mjs <url-or-video-id> [--lang <code>]

Options:
  --lang <code>  Preferred transcript language (default: en)
  --help, -h     Show this help
`;
}

export function parseArgs(argv = process.argv.slice(2)) {
  const args = { url: null, lang: "en", help: false };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--help" || arg === "-h") {
      args.help = true;
    } else if (arg === "--lang") {
      const value = argv[index + 1];
      if (value == null || value.startsWith("--")) throw new Error("--lang requires a value");
      args.lang = value;
      index += 1;
    } else if (!args.url) {
      args.url = arg;
    } else {
      throw new Error(`unrecognized argument: ${arg}`);
    }
  }
  return args;
}

export function main(argv = process.argv.slice(2)) {
  const args = parseArgs(argv);
  if (args.help) {
    console.log(usage());
    return 0;
  }
  if (!args.url) throw new Error("url is required");
  const data = fetchVideo(args.url, args.lang);
  console.log(JSON.stringify(data, null, 2));
  return 0;
}

if (process.argv[1] && realpathSync(process.argv[1]) === fileURLToPath(import.meta.url)) {
  try {
    process.exitCode = main();
  } catch (error) {
    console.error(`ERROR: ${error.message}`);
    process.exitCode = error.exitCode || 1;
  }
}
